### PR TITLE
ipq40xx: add support for MikroTik hAP ac3

### PR DESCRIPTION
**Pull request by robimarko**
This adds support for the MikroTik RouterBOARD RBD53iG-5HacD2HnD
(hAP ac³), a  indoor dual band, dual-radio 802.11ac
wireless AP with external omnidirectional antennae, USB port, five
10/100/1000 Mbps Ethernet ports and PoE passthrough.

See https://mikrotik.com/product/hap_ac3 for more info.

Specifications:
 - SoC: Qualcomm Atheros IPQ4019
 - RAM: 256 MB
 - Storage: 16 MB NOR + 128 MB NAND
 - Wireless:
   · Built-in IPQ4019 (SoC) 802.11b/g/n 2x2:2, 3 dBi antennae
   · Built-in IPQ4019 (SoC) 802.11a/n/ac 2x2:2, 5.5 dBi antennae
 - Ethernet: Built-in IPQ4019 (SoC, QCA8075) , 5x 1000/100/10 port,
             passive PoE in, PoE passtrough on port 5
- 1x USB Type A port

Installation:
1. Boot the initramfs image via TFTP
2. Run "cat /proc/mtd" and look for "ubi" partition mtd device number, ex. "mtd1"
3. Use ubiformat to remove MikroTik specific UBI volumes
* Detach the UBI partition by running: "ubidetach -d 0"
* Format the partition by running: "ubiformat /dev/mtdN -y"
Replace mtdN with the correct mtd index from step 2.
3. Flash the sysupgrade image using "sysupgrade -n"

Signed-off-by: Robert Marko <robimarko@gmail.com>

### CODE REVIEW

**Comment by adschm** [comment]
"image-commands:" -> "build: image:"

File: package/firmware/ipq-wifi/Makefile [hunk 1]
```diff
@@ -44,6 +44,7 @@ ALLWIFIBOARDS:= \
 	linksys_mr8300-v0 \
 	luma_wrtq-329acn \
 	mikrotik_hap-ac2 \
+    mikrotik_hap-ac3 \
```
Comment (adschm) [inline package/firmware/ipq-wifi/Makefile:47]: Broken indent.

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds [hunk 2]
```diff
@@ -45,6 +45,15 @@ engenius,ens620ext)
 	ucidef_set_led_netdev "lan1" "LAN1" "green:lan1" "eth0"
 	ucidef_set_led_netdev "lan2" "LAN2" "green:lan2" "eth1"
 	;;
+mikrotik,hap-ac3)
+	ucidef_set_led_netdev "wan" "WAN" "green:wan" "eth1"
+	ucidef_set_led_netdev "ethernet" "ETHERNET" "green:ethernet" "eth1"
```
Comment (adschm) [inline target/linux/ipq40xx/base-files/etc/board.d/01_leds:50]: ethernet and wan are linked to the same?

File: target/linux/ipq40xx/base-files/etc/board.d/03_gpio_switches [hunk 3]
```diff
@@ -18,6 +18,9 @@ cilab,meshpoint-one)
 compex,wpj428)
 	ucidef_add_gpio_switch "sim_card_select" "SIM card select" "3" "0"
 	;;
+mikrotik,hap-ac3)
+	ucidef_add_gpio_switch "poe_passtrough" "PoE Passthrough" "452" "0"
```
Comment (adschm) [inline target/linux/ipq40xx/base-files/etc/board.d/03_gpio_switches:22]: Same GPIO as in 01_leds?

File: target/linux/ipq40xx/patches-5.10/901-arm-boot-add-dts-files.patch [hunk 4]
```diff
@@ -52,6 +52,7 @@ Signed-off-by: John Crispin <john@phrozen.org>
 +	qcom-ipq4019-fritzbox-7530.dtb \
 +	qcom-ipq4019-fritzrepeater-1200.dtb \
 +	qcom-ipq4019-fritzrepeater-3000.dtb \
++   qcom-ipq4019-hap-ac3.dtb \
```
Comment (adschm) [inline target/linux/ipq40xx/patches-5.10/901-arm-boot-add-dts-files.patch:55]: Broken indent.

File: target/linux/ipq40xx/base-files/etc/board.d/03_gpio_switches (hunk 3 above)
Comment (robimarko) [reply target/linux/ipq40xx/base-files/etc/board.d/03_gpio_switches:22]: No, its just triggered of the PoE GPIO

File: package/firmware/ipq-wifi/Makefile (hunk 1 above)
Comment (robimarko) [reply package/firmware/ipq-wifi/Makefile]: Sorted

File: target/linux/ipq40xx/patches-5.10/901-arm-boot-add-dts-files.patch (hunk 4 above)
Comment (robimarko) [reply target/linux/ipq40xx/patches-5.10/901-arm-boot-add-dts-files.patch]: Sorted

File: scripts/ubinize-image.sh [hunk 5]
```diff
@@ -66,20 +66,22 @@ ubilayout() {
...
+			# squashfs uses 1k block size, ensure we do not
+			# violate that
+			rootsize="$( round_up "$( stat -c%s "$2" )" 1024 )"
+			;;
+		esac
+		ubivol $vol_id rootfs "$2" "$autoresize" "$rootsize"
 
-	vol_id=$(( $vol_id + 1 ))
-	[ "$rootfs_type" = "ubifs" ] || ubivol $vol_id rootfs_data "" 1
+		vol_id=$(( $vol_id + 1 ))
```
[the remark this answers is not in the document]
Comment (robimarko) [reply scripts/ubinize-image.sh:82]: Sorry, I dont get it.
I just added the check for rootfs to this code, other is copy/paste.

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds (hunk 2 above)
Comment (adschm) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:50]: I prefer sticking to the vendor behavior. If people want additional triggers based on preference, they should add them themselves.

Comment (robimarko) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:50]: I got no idea how MikroTik triggers this LED at all.

Comment (adschm) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:50]: So you just assigned the LEDs based on what you thought would be good?

Comment (adschm) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:50]: I don't get it? You are setting an ethernet LED in your code? And internet LED is WAN LED? Why not just use the proper names so there is less confusion?

Comment (adschm) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:50]: > "ethernet" in ROS, so I kept that name and triggered it of the WAN as well.

And that's totally arbitrary, so this should be removed.

> In ROS they have some kind of internet detection, probably some kind of ping test to their server based on which that LED is triggered of.

So OEM does not signal pure WAN link state, like it does for LAN ports?

Comment (robimarko) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:50]: > > "ethernet" in ROS, so I kept that name and triggered it of the WAN as well.
> 
> And that's totally arbitrary, so this should be removed.

Ok, fine will drop it.
> 
> > In ROS they have some kind of internet detection, probably some kind of ping test to their server based on which that LED is triggered of.
> 
> So OEM does not signal pure WAN link-state, like it does for LAN ports?
They do via the port LED and then they also signal internet reachability via the ethernet LED

**Review by voip2014 — COMMENTED** [review]
The changes are working on the latest commit and its functions as expected

**Comment by robimarko** [comment]
@adschm Can we get this merged?
It's been tested thoroughly

**Comment by lukasmrtvy** [comment]
Do We need to wait for major release, or is possible to cherry pick it to v21.02.2 ?

**Comment by garis** [comment]
Tested-by: Riccardo Gasparini garis94@gmail.com

**Comment by robimarko** [comment]
@chunkeey Any chance of you merging this as it's been thoroughly tested?
I will just rebase it to get rid of the conflicts

**Comment by TrashPandaCodingGarbage** [comment]
@chunkeey Please merge. Much love <3

**Comment by tomkins** [comment]
`Tested-by: Alex Tomkins <tomkins@darkzone.net>`

**Review by rolocampusano — APPROVED** [review]

**Comment by hauke** [comment]
Thank you for the patches, I applied them to master in da3261e57cf33d6804222c614c77e13fdb32cf95, 5a305e429facce7fae71f5589b4adbb55b918c9c, 6db4a0372a726ebc1720adf308d698a5df9f1af1, 1fbc9c5e4dea7311236c8f74d6093bc968a710da, f2c4064ecb76bca69d4ce3fbc52209992f2334dc and 3ad229db0b0ed8538eadcdc2a1c3faf964f239be

**Pull request closed without merging**